File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> visited(n);
        vector<vector<int>> graph(n);
        for(auto it: edges){
            graph[it[0]].push_back(it[1]);
            graph[it[1]].push_back(it[0]);
        }
        int countComponents = 0;
        queue<int> q;
        for(int i = 0; i < n; i++){
            if(visited[i] == 0){
                countComponents++;
                q.push(i);
                while(!q.empty()){
                    int node = q.front(); q.pop();
                    for(auto adjNode: graph[node]){
                        if(visited[adjNode] == 0){
                            visited[adjNode] = 1;
                            q.push(adjNode);
                        }
                    }
                }
            }
        }
        return countComponents;
    }
};

File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = []
        for i in range(n):
            graph.append([])

        for edge in edges:
            u = edge[0]
            v = edge[1]
            graph[u].append(v)
            graph[v].append(u)
            
        visited = [0] * n
        
        componentCount = 0
        for i in range(n):
            if visited[i] == 0:
                componentCount += 1

                dq = deque()
                dq.append(i)
                visited[i] = 1

                while len(dq):
                    node = dq.popleft()
                    #now traverse
                    for adjNode in graph[node]:
                        if visited[adjNode] == 0:
                            dq.append(adjNode)
                            visited[adjNode] = 1
        
        return componentCount
                    


